accord-core/src/main/java/accord/primitives/Deps.java

```diff
@@ -0,0 +1,978 @@
+        return new Keys(result);
+    }
+
+    private void ensureTxnIdToKey()
+    {
+        if (txnIdToKey != null)
+            return;
+
+        int[] src = keyToTxnId;
+        int[] trg = txnIdToKey = new int[txnIds.length - keys.size() + keyToTxnId.length];
+
+        // first pass, count number of txnId per key
+        for (int i = keys.size() ; i < src.length ; ++i)
+            trg[src[i]]++;
+
+        // turn into offsets (i.e. add txnIds.size() and then sum them)
+        trg[0] += txnIds.length;
+        for (int i = 1; i < txnIds.length ; ++i)
+            trg[i] += trg[i - 1];
+
+        // shuffle forwards one, so we have the start index rather than end
+        System.arraycopy(trg, 0, trg, 1, txnIds.length - 1);
+        trg[0] = txnIds.length;
+
+        int k = 0;
+        for (int i = keys.size() ; i < src.length ; ++i)
+        {
+            while (i == keyToTxnId[k])
+                ++k;
+
+            trg[trg[src[i]]++] = k;
+        }
+    }
+
+    public void forEachOn(KeyRanges ranges, Predicate<Key> include, BiConsumer<Key, TxnId> forEach)
+    {
+        keys.foldl(ranges, (index, key, value) -> {
+            if (!include.test(key))
+                return null;
+
+            for (int t = index == 0 ? keys.size() : keyToTxnId[index - 1], end = keyToTxnId[index]; t < end ; ++t)
+            {
+                TxnId txnId = txnIds[keyToTxnId[t]];
+                forEach.accept(key, txnId);
+            }
+            return null;
+        }, null);
+    }
+
+    public void forEachOn(CommandStore commandStore, Timestamp executeAt, Consumer<TxnId> forEach)
+    {
+        KeyRanges ranges = commandStore.ranges().since(executeAt.epoch);
+        if (ranges == null)
+            return;
+
+        // TODO: check inlining
+        forEachOn(ranges, commandStore::hashIntersects, forEach);
+    }
+
+    public void forEachOn(KeyRanges ranges, Predicate<Key> include, Consumer<TxnId> forEach)
+    {
+        for (int offset = 0 ; offset < txnIds.length ; offset += 64)
+        {
+            long bitset = keys.foldl(ranges, (keyIndex, key, off, value) -> {
+                if (!include.test(key))
+                    return value;
+
+                int index = keyIndex == 0 ? keys.size() : keyToTxnId[keyIndex - 1];
+                int end = keyToTxnId[keyIndex];
+                if (off > 0)
+                {
+                    // TODO: interpolation search probably great here
+                    index = Arrays.binarySearch(keyToTxnId, index, end, (int)off);
+                    if (index < 0)
+                        index = -1 - index;
+                }
+
+                while (index < end)
+                {
+                    long next = keyToTxnId[index++] - off;
+                    if (next >= 64)
+                        break;
+                    value |= 1L << next;
+                }
+
+                return value;
+            }, offset, 0, 0xffffffffffffffffL);
```

Comment:
I think `-1L` is best,  with `0xffffffffffffffffL` I have to count all the `f` to know how many bits are set

Reply:
I used to get complaints that -1 wasn't clear enough, which is why I switched to 0xf.... I've always preferred -1. Happy to switch.